plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsWizardFragment.java

```diff
@@ -356,22 +366,81 @@ public void widgetSelected(SelectionEvent e) {
 		};
 	}
 
+	private String substitueVariables(String rawSourcePath) {
+		if (org.apache.commons.lang.StringUtils.isNotBlank(rawSourcePath)) {
+			try {
+				return VariablesPlugin.getDefault().getStringVariableManager().performStringSubstitution(rawSourcePath);
+			} catch (CoreException swalloed) {
+				OpenShiftUIActivator.getDefault().getLogger().logError("Could not interpolate "+rawSourcePath);
+			}
+		}
+		return rawSourcePath;
+	}
+
 	private SelectionAdapter onBrowseWorkspace(final Shell shell) {
 		return new SelectionAdapter() {
 
 			@Override
 			public void widgetSelected(SelectionEvent e) {
-				DirectoryDialog dialog = new DirectoryDialog(shell, SWT.OPEN);
-				dialog.setText("Choose the source path to sync");
-				dialog.setFilterPath(ResourcesPlugin.getWorkspace().getRoot().getLocation().toString());
-				String filepath = dialog.open();
-				if (!StringUtils.isEmpty(filepath)) {
-					model.setSourcePath(filepath);
+				String sourcePath = getWorkspaceRelativePath(model.getSourcePath());
+				ElementTreeSelectionDialog dialog = createWorkspaceFolderDialog(shell, sourcePath);
+				if (dialog.open() == IDialogConstants.OK_ID && dialog.getFirstResult() instanceof IContainer) {
+					String path = ((IContainer)dialog.getFirstResult()).getFullPath().toPortableString();
+					String folderPath = VariablesPlugin.getDefault().getStringVariableManager().generateVariableExpression("workspace_loc", path);
+					model.setSourcePath(folderPath);
 				}
 			}
 		};
 	}
 
+	private String getWorkspaceRelativePath(String sourcePath) {
+		if (org.apache.commons.lang.StringUtils.isBlank(sourcePath) || sourcePath.contains("{")) {
+			return sourcePath;
+		}
+		IPath absolutePath = new Path(sourcePath);
+		IContainer container = ResourcesPlugin.getWorkspace().getRoot().getContainerForLocation(absolutePath);
+		if (container != null) {
+			return container.getFullPath().toPortableString();
+		}
+		return null;
+	}
+
+	private ElementTreeSelectionDialog createWorkspaceFolderDialog(Shell shell, String selectedFile) {
+		ElementTreeSelectionDialog dialog = new ElementTreeSelectionDialog(
+				shell,
+				new WorkbenchLabelProvider(),
+				new WorkbenchContentProvider()
+		);
+		dialog.setTitle("Select a workspace folder");
+		dialog.setMessage("Select a workspace folder to deploy");
+		dialog.setInput( ResourcesPlugin.getWorkspace().getRoot() );
+		dialog.addFilter(new ViewerFilter() {
+
+			@Override
+			public boolean select(Viewer viewer, Object parentElement, Object element) {
+				return element instanceof IContainer;
+			}
+		});
+		dialog.setAllowMultiple( false );
+		org.eclipse.core.resources.IResource res = model.getDeployProject();
+		if (org.apache.commons.lang.StringUtils.isNotBlank(selectedFile)) {
+			String prefix = "${workspace_loc:";
```

Comment:
"workspace_loc" is used 3 times. Maybe it deserves a constant. Then for prefix it may be "${" + WORKSPACE_LOC.
It is a pity that StringSubstitutionEngine.VARIABLE_START is a private constant.